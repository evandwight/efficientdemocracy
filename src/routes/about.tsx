import { reactAboutRender } from '../views/utils';

export const renderAbout = (view, title) => {
    return (req, res) => {
        reactAboutRender(res, view(), { title });
    }
}