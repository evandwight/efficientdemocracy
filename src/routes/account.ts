const bcrypt = require("bcrypt");
import { reactRender } from '../views/utils';
import db from '../db/databaseApi';
import { Login } from '../views/login';
import { Register } from '../views/register';
import { Strikes } from '../views/strikes';
import {UserSettings} from '../views/userSettings';
import * as C from "../constant";
import validator from 'validator';

export const USERNAME_REGEX = RegExp('[A-Za-z0-9_]*');

export function login(req, res) {
    const {errors, csrfToken} = res.locals;
    reactRender(res, Login({errors, csrfToken}), {showLogin: false, title:"Login"});
}

export function register(req, res) {
    const {errors, csrfToken} = res.locals;
    reactRender(res, Register({errors, defaultValues:{}, csrfToken}), {showLogin:false, title:"Register"});
}

export async function userRegister(req, res) {
    let { userName, name, email, password, password2, code } = req.body;
    let errors = [];

    console.log({
        userName,
        name,
        email,
        password,
        password2
    });

    if (code != process.env.LOGIN_CODE){
        errors.push({ message: "Invalid code" });
    }

    if (!userName || !name || !email || !password || !password2) {
        errors.push({ message: "Please enter all fields" });
    }

    if (!USERNAME_REGEX.test(userName)) {
        errors.push({ message: "Username may only contain alphanumeric characters and underscore" })
    }

    if (password.length < 6) {
        errors.push({ message: "Password must be a least 6 characters long" });
    }

    if (password !== password2) {
        errors.push({ message: "Passwords do not match" });
    }

    if (!validator.isEmail(email)) {
        errors.push({ message: "Invalid email" });
    }

    const user = await  db.users.getUserByEmail(email)
    if (user !== null) {
        errors.push({message: "Email already registered"});
    }

    if (errors.length > 0) {
        const defaultValues = {userName, name, email, password, password2, code};
        reactRender(res, Register({errors, defaultValues, csrfToken: req.csrfToken()}), {showLogin: false, title:"Register"});
    } else {
        const hashedPassword = await bcrypt.hash(password, 10);
        // Validation passed
        return db.users.createUser({ userName, name, email, hashedPassword }).then(() => {
            res.redirect(C.URLS.USER_LOGIN);
        });
    }
}

export async function logout(req, res) {
    req.logout();
    res.redirect("/")
}

export async function strikes(req, res) {
    const {user} = res.locals;
    const strikes = await db.modActions.getStrikes(user.id);
    reactRender(res, Strikes({user, strikes}), {title: "Strikes"});
}

export async function userSettings(req, res) {
    const {user} = res.locals;
    reactRender(res, UserSettings({user}), {title: "User settings"});
}