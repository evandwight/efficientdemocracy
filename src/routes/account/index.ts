export * from './userSettings';
export * from './localAccount';
export * from './genericAccount';
export * from './strikes';
export * from './unsubscribe';
export * from './verifyEmailGoodness';


