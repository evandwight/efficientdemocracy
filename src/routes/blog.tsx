import { reactAboutRender } from '../views/utils';
import { BlogPage, BlogEntries } from '../views/blog';
import type { BlogEntry } from '../views/blog/blogEntry';
import { rssXmlBoilerPlate } from '../views/blog/rss';

// TODO https://guides.github.com/features/mastering-markdown/#what
// https://www.npmjs.com/package/markdown-it
// put entries in the database and load them

export const renderBlog = (page: BlogEntry) => {
    return (req, res) => {
        reactAboutRender(res, page.element(), { title: page.title, includeScript: page.includeScript, includeBlogRssHint: true });
    }
}

let blogEntries = Object.values(BlogEntries);
blogEntries.sort((a, b) => b.created.getTime() - a.created.getTime());
const cachedRssXml = rssXmlBoilerPlate(blogEntries);

export const index = (req, res) => {
    const page = req.params.page ? parseInt(req.params.page) : 0;
    const entriesPerPage = 10;

    const entries = blogEntries.map(blog => ({ title: blog.title, url: blog.url, created: blog.created }));

    const pageEntries = entries.slice(page * entriesPerPage, (page + 1) * entriesPerPage);

    const showNext = ((page + 1) * entriesPerPage) < entries.length;
    const title = (page > 0 ? "Blog " + page : "Blog");
    reactAboutRender(res, BlogPage({ entries: pageEntries, page, showNext }), { title, includeBlogRssHint: true });
}

export const getRssXml = (req, res) => {
    res.setHeader('content-type', 'application/atom+xml');
    res.send(cachedRssXml);
}