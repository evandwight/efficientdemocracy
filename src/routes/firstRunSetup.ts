import { reactRender } from '../views/utils';
import db from '../db/databaseApi';
import {FirstRunSetup} from '../views/firstRunSetup';
import * as C from '../constant';

export const USERNAME_REGEX = RegExp('[A-Za-z0-9_]*');

export function firstRunSetup(req, res) {
    const {csrfToken} = res.locals;
    reactRender(res, FirstRunSetup({csrfToken}), {showLogin: false, title:"Account setup"});
}

export async function submitFirstRunSetup(req, res) {
    const userId = req.user.id;
    const body = { ...req.body };
    const sendEmails = body.hasOwnProperty("send_emails");
    
    await db.users.setSendEmails({userId, sendEmails});
    await db.users.setFirstRunComplete(userId);

    res.redirect(C.URLS.QPOSTS);
}