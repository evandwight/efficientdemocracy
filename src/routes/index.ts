export * as Mods from './mods';
export * as Account from './account';
export * as QPost from './qPost';
export * as About from './about';
export * as Blog from './blog';
export * as Sample from './sample';
export * as Misc from './misc';
export * as FirstRunSetup from './firstRunSetup';