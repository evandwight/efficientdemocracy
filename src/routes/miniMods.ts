import db from '../db/databaseApi';
import * as C from '../constant';
import { reactRender } from '../views/utils';
import { PostModActions } from '../views/miniMods/postModActions';
import validator from 'validator';
import DemocraticModerationService from '../services/democraticModerationService';
import { formToStrikes, validationAssert, ValidationError } from './utils';

export async function viewPostActions(req, res) {
    const { id } = req.params;
    const userId = req.user.id;
    validationAssert(validator.isUUID(id, 4), "Invalid thing id", 400);

    const thingType = await db.things.getType(id);
    validationAssert(!!thingType, "Thing not found", 404);
    validationAssert(C.THINGS.QPOST === thingType, "Invalid thing type", 400);
    
    const post = await db.qPosts.getPost(id);
    validationAssert(!!post, "Thing not found", 400);
    const miniModVotes = await DemocraticModerationService.getMiniModVotes({userId, thingId: id});
    
    reactRender(res, PostModActions({ post, miniModVotes, csrfToken: res.locals.csrfToken }), { title: "Mini mod actions" });
}



export async function submitPostAction(req, res) {
    const thingId = req.params.id;
    validationAssert(validator.isUUID(thingId, 4), "Invalid thing id", 400);
    const { field } = req.params;
    const userId = req.user.id;
    const body = { ...req.body };
    const vote = body.hasOwnProperty("value");
    const strikes = formToStrikes(body, {disallowStrikeDisputers: true});

    await DemocraticModerationService.miniModVote({thingId, field, userId, vote, ... strikes});

    res.redirect(req.get("Referrer"));
}