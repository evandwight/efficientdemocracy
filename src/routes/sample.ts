import db from '../db/databaseApi';
import * as C from '../constant';
import { reactRender } from '../views/utils';
import assert from 'assert';
import validator from 'validator';
import { ViewSampleResult } from '../views/viewSampleResult';
import { sum } from '../batch/samples';
import { ViewSamples } from '../views/viewSamples';

export async function submitSampleVote(req, res) {
    const {sampleId} = req.params;
    assert(validator.isUUID(sampleId, 4), "Invalid sampleId");
    const userId = req.user.id;
    const body = {... req.body};
    const [vote, strikeUps, strikeDowns, strikePoster, strikeDisputers] = 
        ["vote", "strike_ups", "strike_downs", "strike_poster", "strike_disputers"].map((v) => body.hasOwnProperty(v));

    assert(await db.samples.canUserVote({sampleId, userId}), "User not in sample");

    const {field} = await db.samples.getSampleResult(sampleId);
    if (field === C.SAMPLE.FIELDS.DEEPLY_IMPORTANT) {
        assert(!(strikeUps || strikeDowns || strikePoster), "Cannot strike with field deeply important");
    }

    db.samples.vote({sampleId, userId, vote, strikeUps, strikeDowns, strikePoster, strikeDisputers})

    res.redirect(req.get("Referrer"));
}

export function countsToDataTable(counts) {
    const rows: any[] = [true, false, null];
    const rowHeader = ["Yes", "No", "Didn't vote"]
    const cols = ["strike_ups", "strike_downs", "strike_poster", "strike_disputers"];

    return [
        ["Censor?", "Total", "Strike up voters", "Strike down voters", "Strike poster", "Strike disputers"],
        ... rows.map((row, i) => [
            rowHeader[i],
            sum(counts.filter(v => v.vote === row)), // Total
            ... cols.map(col => sum(counts.filter(v => v.vote === row && v[col] === true)))
            ])
        ];
}

export function countsToChartData(sample) {
    const {counts} = sample;
    let chartData: any = {
        shouldCensor: {
            labels: ["Yes", "No", "Didn't vote"],
            datasets: [{
                label: '# of Votes',
                data: [true, false, null].map(v => sum(counts.filter(e => e.vote === v))),
            }]
        }
    };
    if (!!sample.result) {
        chartData.shouldStrike = {
            labels: ["Up voters", "Down voters", "Poster", "Disputers"],
            datasets: [{
                label: '# of Votes',
                data: ["strike_ups", "strike_downs", "strike_poster", "strike_disputers"].map(v =>
                    sum(counts.filter(e => e.vote === sample.result.vote && e[v] === true))),
            }]
        };
    }
    return chartData;
}

export async function viewSampleResult(req, res) {
    const {sampleId} = req.params;
    assert(validator.isUUID(sampleId, 4));

    const sample = await db.samples.getSampleResult(sampleId);
    assert(sample.is_complete, "Sample must be complete to view");

    const dangerousChartData = countsToChartData(sample);
    const dataTable = countsToDataTable(sample.counts);
    
    reactRender(res, ViewSampleResult({ sample, dataTable }), {title: "Sample", includeChartJs: true, dangerousChartData});
}

export async function viewSamples(req, res) {
    const {thingId} = req.params;
    assert(validator.isUUID(thingId, 4));

    const samples = await db.samples.getCompletedSamples(thingId);
    
    reactRender(res, ViewSamples({ samples }), {title: "Samples"});
}