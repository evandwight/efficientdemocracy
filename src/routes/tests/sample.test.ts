import { countsToDataTable, countsToChartData } from '../../routes/sample';


describe('countsToDataTable', () => {
    it('works', () => {
        const counts = [
            {vote: true, strike_ups: false, strike_downs: false, strike_poster: false, strike_disputers: false, count: 1},
            {vote: false, strike_ups: true, strike_downs: true, strike_poster: true, strike_disputers: true, count: 2},
            {vote: null, strike_ups: true, strike_downs: true, strike_poster: true, strike_disputers: true, count: 3},
        ]
        const res = countsToDataTable(counts);
        expect(res).toMatchSnapshot();
    });
});

describe('countsToChartData', () => {
    it('works', () => {
        const counts = [
            {vote: true, strike_ups: false, strike_downs: false, strike_poster: false, strike_disputers: false, count: 1},
            {vote: false, strike_ups: true, strike_downs: true, strike_poster: true, strike_disputers: true, count: 2},
            {vote: null, strike_ups: true, strike_downs: true, strike_poster: true, strike_disputers: true, count: 3},
        ]
        const sample = {
            result: {vote: false, strike_ups: true, strike_downs: true, strike_poster: true, strike_disputers: true},
            counts
        }
        const res = countsToChartData(sample);
        expect(res).toMatchSnapshot();
    });
    it('handles null results', () => {
        const counts = [
            {vote: true, strike_ups: false, strike_downs: false, strike_poster: false, strike_disputers: false, count: 1},
            {vote: false, strike_ups: true, strike_downs: true, strike_poster: true, strike_disputers: true, count: 2},
            {vote: null, strike_ups: true, strike_downs: true, strike_poster: true, strike_disputers: true, count: 3},
        ]
        const sample = {
            result: null,
            counts
        }
        const res = countsToChartData(sample);
        expect(res).toMatchSnapshot();
    });
});