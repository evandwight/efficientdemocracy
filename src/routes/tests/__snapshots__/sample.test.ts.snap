// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`countsToChartData handles null results 1`] = `
Object {
  "shouldCensor": Object {
    "datasets": Array [
      Object {
        "data": Array [
          1,
          2,
          3,
        ],
        "label": "# of Votes",
      },
    ],
    "labels": Array [
      "Yes",
      "No",
      "Didn't vote",
    ],
  },
}
`;

exports[`countsToChartData works 1`] = `
Object {
  "shouldCensor": Object {
    "datasets": Array [
      Object {
        "data": Array [
          1,
          2,
          3,
        ],
        "label": "# of Votes",
      },
    ],
    "labels": Array [
      "Yes",
      "No",
      "Didn't vote",
    ],
  },
  "shouldStrike": Object {
    "datasets": Array [
      Object {
        "data": Array [
          2,
          2,
          2,
          2,
        ],
        "label": "# of Votes",
      },
    ],
    "labels": Array [
      "Up voters",
      "Down voters",
      "Poster",
      "Disputers",
    ],
  },
}
`;

exports[`countsToDataTable works 1`] = `
Array [
  Array [
    "Censor?",
    "Total",
    "Strike up voters",
    "Strike down voters",
    "Strike poster",
    "Strike disputers",
  ],
  Array [
    "Yes",
    1,
    0,
    0,
    0,
    0,
  ],
  Array [
    "No",
    2,
    2,
    2,
    2,
    2,
  ],
  Array [
    "Didn't vote",
    3,
    3,
    3,
    3,
    3,
  ],
]
`;
